
### 1. Importing the libraries we need and reading the dataset into pandas.


In [83]:
import pandas as pd
import numpy as np

### 2. Loading Dataset into our Jupyter Notebook


In [84]:
autos = pd.read_csv("autos_clean.csv", encoding="Latin-1")

autos.info()
autos.describe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Unnamed: 0           30000 non-null  int64 
 1   dateCrawled          30000 non-null  object
 2   name                 30000 non-null  object
 3   seller               30000 non-null  object
 4   offerType            30000 non-null  object
 5   price                30000 non-null  int64 
 6   abtest               30000 non-null  object
 7   vehicleType          26963 non-null  object
 8   yearOfRegistration   30000 non-null  int64 
 9   gearbox              28420 non-null  object
 10  powerPS              30000 non-null  int64 
 11  model                28320 non-null  object
 12  kilometer            30000 non-null  int64 
 13  monthOfRegistration  30000 non-null  int64 
 14  fuelType             27330 non-null  object
 15  brand                30000 non-null  object
 16  notR

<bound method NDFrame.describe of        Unnamed: 0          dateCrawled  \
0           52489  2016-03-16 21:37:36   
1          174423  2016-04-01 13:36:58   
2           91195  2016-04-03 15:45:31   
3          138395  2016-03-30 11:52:34   
4          190527  2016-04-01 16:37:30   
...           ...                  ...   
29995      124522  2016-03-31 19:46:09   
29996       38277  2016-03-23 19:45:50   
29997      147623  2016-03-24 10:58:42   
29998      213272  2016-03-09 16:55:55   
29999       43445  2016-04-04 20:44:18   

                                                    name  seller offerType  \
0                   Chrysler_300_M_3.5__TÜV_neu_03/2018  privat   Angebot   
1                  Renault_Clio_Bastlerfahrzeug_mit_TÜV  privat   Angebot   
2                                      Auto_zu_verkaufen  privat   Angebot   
3                             Opel_astra_g_Coupe_bertone  privat   Angebot   
4                                     Seat_Ibiza_1_9_TDI  privat   Ange

In [85]:
autos.head(5)

Unnamed: 0          dateCrawled                                   name  \
0       52489  2016-03-16 21:37:36   Chrysler_300_M_3.5__TÜV_neu_03/2018   
1      174423  2016-04-01 13:36:58  Renault_Clio_Bastlerfahrzeug_mit_TÜV   
2       91195  2016-04-03 15:45:31                      Auto_zu_verkaufen   
3      138395  2016-03-30 11:52:34             Opel_astra_g_Coupe_bertone   
4      190527  2016-04-01 16:37:30                     Seat_Ibiza_1_9_TDI   

   seller offerType  price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot   4900  control   limousine                2004   
1  privat   Angebot    300     test  kleinwagen                1996   
2  privat   Angebot    400     test   limousine                2001   
3  privat   Angebot   2800     test       coupe                2004   
4  privat   Angebot    950  control  kleinwagen                2000   

     gearbox  ...    model kilometer  monthOfRegistration  fuelType     brand  \
0  automatik  ...   andere    150000                    3    benzin  chrysler   
1    manuell  ...     clio    150000                   11    benzin   renault   
2    manuell  ...  primera      5000                    0    benzin    nissan   
3    manuell  ...    astra    150000                    0    benzin      opel   
4    manuell  ...    ibiza    150000                   10    diesel      seat   

  notRepairedDamage          dateCreated nrOfPictures  postalCode  \
0              nein  2016-03-16 00:00:00            0       88069   
1               NaN  2016-04-01 00:00:00            0       33335   
2                ja  2016-04-03 00:00:00            0       13627   
3               NaN  2016-03-30 00:00:00            0       25832   
4                ja  2016-04-01 00:00:00            0       53925   

              lastSeen  
0  2016-04-07 13:15:40  
1  2016-04-07 11:18:05  
2  2016-04-05 14:46:21  
3  2016-04-07 02:45:04  
4  2016-04-01 16:37:30  

[5 rows x 21 columns]

### 3. Cleaning Column Names

In [86]:
del autos['Unnamed: 0']

In [87]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [88]:
# Changing column names from camelcase to snakecase
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [89]:
autos.head(5)

date_crawled                                   name  seller  \
0  2016-03-16 21:37:36   Chrysler_300_M_3.5__TÜV_neu_03/2018  privat   
1  2016-04-01 13:36:58  Renault_Clio_Bastlerfahrzeug_mit_TÜV  privat   
2  2016-04-03 15:45:31                      Auto_zu_verkaufen  privat   
3  2016-03-30 11:52:34             Opel_astra_g_Coupe_bertone  privat   
4  2016-04-01 16:37:30                     Seat_Ibiza_1_9_TDI  privat   

  offer_type  price  ab_test vehicle_type  registration_year    gearbox  \
0    Angebot   4900  control    limousine               2004  automatik   
1    Angebot    300     test   kleinwagen               1996    manuell   
2    Angebot    400     test    limousine               2001    manuell   
3    Angebot   2800     test        coupe               2004    manuell   
4    Angebot    950  control   kleinwagen               2000    manuell   

   power_ps    model  odometer  registration_month fuel_type     brand  \
0       252   andere    150000                   3    benzin  chrysler   
1         0     clio    150000                  11    benzin   renault   
2       120  primera      5000                   0    benzin    nissan   
3       147    astra    150000                   0    benzin      opel   
4        90    ibiza    150000                  10    diesel      seat   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-16 00:00:00           0        88069   
1               NaN  2016-04-01 00:00:00           0        33335   
2                ja  2016-04-03 00:00:00           0        13627   
3               NaN  2016-03-30 00:00:00           0        25832   
4                ja  2016-04-01 00:00:00           0        53925   

             last_seen  
0  2016-04-07 13:15:40  
1  2016-04-07 11:18:05  
2  2016-04-05 14:46:21  
3  2016-04-07 02:45:04  
4  2016-04-01 16:37:30


### 4. Initial Exploration and Cleaning.

Lets explore the data to find obvious areas where we can clean the data.


In [90]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 30000        30000   30000      30000  3.000000e+04   
unique                28285        23511       1          2           NaN   
top     2016-03-08 19:44:17  Ford_Fiesta  privat    Angebot           NaN   
freq                      4           71   30000      29998           NaN   
mean                    NaN          NaN     NaN        NaN  1.058453e+04   
std                     NaN          NaN     NaN        NaN  5.924220e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.000000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       ab_test vehicle_type  registration_year  gearbox      power_ps  model  \
count    30000        26963       30000.000000    28420  30000.000000  28320   
unique       2            8                NaN        2           NaN    241   
top       test    limousine                NaN  manuell           NaN   golf   
freq     15597         7644                NaN    22308           NaN   2392   
mean       NaN          NaN        2005.361367      NaN    115.719833    NaN   
std        NaN          NaN         114.054959      NaN    222.231276    NaN   
min        NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%        NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%        NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%        NaN          NaN        2008.000000      NaN    148.000000    NaN   
max        NaN          NaN        9999.000000      NaN  19312.000000    NaN   

             odometer  registration_month fuel_type       brand  \
count    30000.000000        30000.000000     27330       30000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     18161        6385   
mean    125729.500000            5.732933       NaN         NaN   
std      39857.746438            3.697134       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count              24077                30000     30000.0  30000.000000   
unique                 2                   58         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq               21208                 1189         NaN           NaN   
mean                 NaN                  NaN         0.0  51007.737833   
std                  NaN                  NaN         0.0  25899.208456   
min                  NaN                  NaN         0.0   1067.000000   
25%                  NaN                  NaN         0.0  30459.000000   
50%                  NaN                  NaN         0.0  49661.000000   
75%                  NaN                  NaN         0.0  72108.000000   
max                  NaN                  NaN         0.0  99998.000000   

                  last_seen  
count                 30000  
unique                24924  
top     2016-04-07 03:46:06  
freq                      6  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Observations:
    - There are columns with text where all values are the same
            1. seller
            2. offer_type
    - num_photos column is odd, and taking a value count will check the values stored

In [91]:
autos['num_photos'].value_counts()

# The num_photos column has 0 for every column.
# Dropping the column because it serves no purpose and unhelpful in our analysis

autos = autos.drop(['num_photos'], axis=1)

In [92]:
# Also we drop the column and the seller and offer_type
autos = autos.drop(['seller', 'offer_type'], axis=1)

In [93]:
# Next we convert the 'price' and 'odometer' because there are characters that are stored as text instead of numerical values

autos['price'] = (autos['price'].astype(str)
                            .str.replace("$","")
                            .str.replace(",","")
                            .astype(int)
                            )
autos['price'].head()

C:\Users\kuria\AppData\Local\Temp\ipykernel_4124\404908277.py:4: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  .str.replace("$","")


0    4900
1     300
2     400
3    2800
4     950
Name: price, dtype: int32

In [94]:
autos['odometer'] = (autos['odometer'].astype(str)
                            .str.replace("km", "")
                            .str.replace(",", "")
                            .astype(int)
                    )
# renaming odometer to odometer_km
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
autos["odometer_km"].head()

0    150000
1    150000
2      5000
3    150000
4    150000
Name: odometer_km, dtype: int32

### 5. Exploring the Odometer and Price Columns


In [95]:
autos['odometer_km'].value_counts()

150000    19438
125000     3076
100000     1268
90000      1078
80000       861
70000       796
60000       728
50000       627
5000        533
40000       515
30000       480
20000       457
10000       143
Name: odometer_km, dtype: int64

It seems the values are already rounded indicating the sellers had chose from ore-set options. In addition, there are more high mileage than low mileage vehicles.

In [96]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
autos['price'].value_counts().head(10)

(1881,)
count    3.000000e+04
mean     1.058453e+04
std      5.924220e+05
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.000000e+03
max      1.000000e+08
Name: price, dtype: float64


0       829
1500    466
500     440
1200    399
2500    398
1000    363
3500    320
600     308
800     289
999     273
Name: price, dtype: int64

The prices in the column are also rounded with 1887 unique values, this are rounded prices from the site.

There are 829 cars listed with $0 price - we might consider removing these rows.

Lastly, the max price is one hundred million dollars, lets take a look at the highest orices further.

In [97]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
12345678    2
11111111    1
9999999     1
650000      1
300000      1
270000      1
250000      1
219000      1
198000      1
197500      1
195900      1
164990      1
148000      1
137900      1
129900      1
123456      1
120911      1
120000      1
119911      1
Name: price, dtype: int64

In [98]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     829
1      96
5       2
9       1
10      7
11      2
12      1
13      1
14      1
15      3
18      1
19      1
20      1
25      3
30      4
35      2
40      1
49      1
50     29
55      4
Name: price, dtype: int64

Lets remove anything above $350000 as the prices are steadily increasing to that number and then jump to less realistic numbers.

In [99]:
autos = autos[autos['price'].between(1, 351000)]
autos['price'].describe()

count     29165.000000
mean       5866.054860
std        8706.673161
min           1.000000
25%        1250.000000
50%        3000.000000
75%        7400.000000
max      300000.000000
Name: price, dtype: float64

### 6. Exploring the date columns


There are 5 columns representing date values;
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.


Lets exlore each off the columns to learn more about the listings.

In [100]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-16 21:37:36  2016-03-16 00:00:00  2016-04-07 13:15:40
1  2016-04-01 13:36:58  2016-04-01 00:00:00  2016-04-07 11:18:05
2  2016-04-03 15:45:31  2016-04-03 00:00:00  2016-04-05 14:46:21
3  2016-03-30 11:52:34  2016-03-30 00:00:00  2016-04-07 02:45:04
4  2016-04-01 16:37:30  2016-04-01 00:00:00  2016-04-01 16:37:30

Lets check missing values in the distribution and use percentages iinstead of counts.

In [101]:
(autos['date_crawled']
                .str[:10]
                .value_counts(normalize=True, dropna=False)
                .sort_index()
)


2016-03-05    0.025716
2016-03-06    0.013544
2016-03-07    0.036996
2016-03-08    0.032779
2016-03-09    0.034048
2016-03-10    0.032436
2016-03-11    0.034391
2016-03-12    0.038162
2016-03-13    0.014572
2016-03-14    0.036105
2016-03-15    0.032676
2016-03-16    0.029385
2016-03-17    0.031339
2016-03-18    0.011692
2016-03-19    0.036071
2016-03-20    0.035419
2016-03-21    0.036722
2016-03-22    0.029419
2016-03-23    0.032539
2016-03-24    0.029865
2016-03-25    0.030276
2016-03-26    0.032950
2016-03-27    0.029385
2016-03-28    0.032779
2016-03-29    0.034048
2016-03-30    0.033533
2016-03-31    0.035419
2016-04-01    0.036173
2016-04-02    0.035419
2016-04-03    0.039465
2016-04-04    0.038608
2016-04-05    0.013544
2016-04-06    0.003086
2016-04-07    0.001440
Name: date_crawled, dtype: float64

In [104]:
(autos['date_crawled']
                .str[:10]
                .value_counts(normalize=True, dropna=False)
                .sort_values()
)


2016-04-07    0.001440
2016-04-06    0.003086
2016-03-18    0.011692
2016-03-06    0.013544
2016-04-05    0.013544
2016-03-13    0.014572
2016-03-05    0.025716
2016-03-27    0.029385
2016-03-16    0.029385
2016-03-22    0.029419
2016-03-24    0.029865
2016-03-25    0.030276
2016-03-17    0.031339
2016-03-10    0.032436
2016-03-23    0.032539
2016-03-15    0.032676
2016-03-28    0.032779
2016-03-08    0.032779
2016-03-26    0.032950
2016-03-30    0.033533
2016-03-09    0.034048
2016-03-29    0.034048
2016-03-11    0.034391
2016-03-20    0.035419
2016-04-02    0.035419
2016-03-31    0.035419
2016-03-19    0.036071
2016-03-14    0.036105
2016-04-01    0.036173
2016-03-21    0.036722
2016-03-07    0.036996
2016-03-12    0.038162
2016-04-04    0.038608
2016-04-03    0.039465
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [105]:
(autos['last_seen']
                .str[:10]
                .value_counts(normalize=True, dropna=False)
                .sort_index()
)


2016-03-05    0.001234
2016-03-06    0.003840
2016-03-07    0.005417
2016-03-08    0.007920
2016-03-09    0.010183
2016-03-10    0.010355
2016-03-11    0.013269
2016-03-12    0.025647
2016-03-13    0.008538
2016-03-14    0.012241
2016-03-15    0.016767
2016-03-16    0.015704
2016-03-17    0.027122
2016-03-18    0.006378
2016-03-19    0.017144
2016-03-20    0.020710
2016-03-21    0.019990
2016-03-22    0.020127
2016-03-23    0.018893
2016-03-24    0.019030
2016-03-25    0.018824
2016-03-26    0.017521
2016-03-27    0.016390
2016-03-28    0.020778
2016-03-29    0.022630
2016-03-30    0.024447
2016-03-31    0.023384
2016-04-01    0.023007
2016-04-02    0.026882
2016-04-03    0.025853
2016-04-04    0.024824
2016-04-05    0.124670
2016-04-06    0.218310
2016-04-07    0.131973
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which asllows us to determine on what day a listing was removeed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are ore than the calues from previous days, its likely there was massive spike in sales, and more likely that these values are to do with the crawling period ending and dont necessary indicate car sales.

In [106]:
print(autos['ad_created'].str[:10].unique().shape)
(autos['ad_created']
    .str[:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index()
    )

(58,)


2015-09-04    0.000034
2016-01-13    0.000034
2016-01-15    0.000034
2016-01-23    0.000034
2016-01-25    0.000034
2016-02-01    0.000034
2016-02-10    0.000034
2016-02-11    0.000034
2016-02-14    0.000034
2016-02-17    0.000034
2016-02-18    0.000069
2016-02-19    0.000034
2016-02-20    0.000069
2016-02-21    0.000103
2016-02-22    0.000034
2016-02-24    0.000034
2016-02-25    0.000034
2016-02-26    0.000034
2016-02-28    0.000206
2016-02-29    0.000240
2016-03-01    0.000137
2016-03-02    0.000103
2016-03-03    0.001200
2016-03-04    0.001852
2016-03-05    0.023007
2016-03-06    0.014469
2016-03-07    0.035693
2016-03-08    0.033190
2016-03-09    0.033705
2016-03-10    0.032265
2016-03-11    0.034802
2016-03-12    0.038059
2016-03-13    0.015704
2016-03-14    0.034493
2016-03-15    0.033362
2016-03-16    0.029316
2016-03-17    0.031030
2016-03-18    0.012755
2016-03-19    0.035042
2016-03-20    0.035865
2016-03-21    0.036722
2016-03-22    0.028802
2016-03-23    0.032608
2016-03-24 

There is large variety of ad created dates.Most fall within a month of the listing date, but a few are quite old, with the oldest at around 9 months.

In [107]:
autos['registration_year'].describe()

count    29165.000000
mean      2004.748431
std         82.223946
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year tthat the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values, The max value is 100, long before cars were invented and the max is 9999, many yaers into the future.

### 7. Dealing with Incorrect Registration Year Data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.#

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [108]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.03953368763929367

Given that this is less than 4% of our data, we will remove these rows.

In [109]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.064972
1999    0.064401
2005    0.062259
2004    0.058118
2003    0.057547
2006    0.057368
2001    0.057011
2002    0.055262
1998    0.050300
2007    0.050086
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

### 8. Exploring Price by Brand

In [111]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.209553
opel              0.109489
bmw               0.109346
mercedes_benz     0.093924
audi              0.089712
ford              0.069756
renault           0.046908
peugeot           0.029916
fiat              0.025596
seat              0.017207
skoda             0.016243
mazda             0.015493
citroen           0.015101
smart             0.013566
nissan            0.012959
toyota            0.012423
sonstige_autos    0.009853
mini              0.009639
hyundai           0.009639
mitsubishi        0.009532
volvo             0.008746
kia               0.007854
honda             0.007675
alfa_romeo        0.007140
suzuki            0.007033
chevrolet         0.005962
porsche           0.005712
chrysler          0.003998
land_rover        0.002570
dacia             0.002535
subaru            0.002178
daihatsu          0.001999
jaguar            0.001928
jeep              0.001821
daewoo            0.001714
saab              0.001464
lancia            0.001285
t

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [116]:
brand_counts = autos['brand'].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [117]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price'].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5420,
 'opel': 2987,
 'bmw': 8525,
 'mercedes_benz': 8753,
 'audi': 9077,
 'ford': 3554}

Of the top 5 brands, there is a distinct price gap:

- Audi, BMW and Mercedes Benz are of higher cost
-  Ford and Opel are less costly
- And, Volkswagen is in the average cost wise

### 9. Storing Aggregate Data in a DataFrame


In [119]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=['mean_price'])

mean_price
volkswagen           5420
opel                 2987
bmw                  8525
mercedes_benz        8753
audi                 9077
ford                 3554

In [121]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)


In [122]:
brand_info = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  133072
opel                 129701
audi                 129194
mercedes_benz        128954
volkswagen           128586
ford                 123697

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.